File: Habitica/src/main/java/com/habitrpg/android/habitica/data/UserRepository.java
package com.habitrpg.android.habitica.data;

import android.support.annotation.Nullable;
import android.support.annotation.StringDef;

import com.habitrpg.android.habitica.models.Skill;
import com.habitrpg.android.habitica.models.inventory.Customization;
import com.habitrpg.android.habitica.models.inventory.CustomizationSet;
import com.habitrpg.android.habitica.models.responses.SkillResponse;
import com.habitrpg.android.habitica.models.responses.UnlockResponse;
import com.habitrpg.android.habitica.models.social.ChatMessage;
import com.habitrpg.android.habitica.models.tasks.Task;
import com.habitrpg.android.habitica.models.user.Stats;
import com.habitrpg.android.habitica.models.user.User;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Map;

import io.realm.RealmResults;
import rx.Observable;

public interface UserRepository extends BaseRepository {

    Observable<User> getUser(String userID);
    Observable<User> getUser();
    Observable<User> updateUser(@Nullable User user, Map<String, Object> updateData);
    Observable<User> updateUser(@Nullable User user, String key, Object value);

    Observable<User> retrieveUser(Boolean withTasks);
    Observable<User> retrieveUser(Boolean withTasks, Boolean forced);

    Observable<RealmResults<ChatMessage>> getInboxMessages(String replyToUserID);
    Observable<RealmResults<ChatMessage>> getInboxOverviewList();

    Observable<User> revive(User user);

    void resetTutorial(@Nullable User user);

    Observable<User> sleep(User user);

    Observable<RealmResults<Skill>> getSkills(User user);

    Observable<RealmResults<Skill>> getSpecialItems(User user);

    Observable<SkillResponse> useSkill(@Nullable User user, String key, String target, String taskId);

    Observable<SkillResponse> useSkill(@Nullable User user, String key, String target);

    Observable<User> changeClass();

    Observable<User> disableClasses();

    Observable<User> changeClass(String selectedClass);

    Observable<UnlockResponse> unlockPath(User user, Customization customization);
    Observable<UnlockResponse> unlockPath(User user, CustomizationSet set);

    void runCron(List<Task> tasks);
    void runCron();

    Observable<List> readNotification(String id);

    Observable<User> changeCustomDayStart(int dayStartTime);

    Observable<User> updateLanguage(User user, String languageCode);

    Observable<User> resetAccount();
    Observable<Void> deleteAccount(String password);

    Observable<Void> sendPasswordResetEmail(String email);

    Observable<Void> updateLoginName(@NotNull String newLoginName, @NotNull String password);
    Observable<Void> updateEmail(@NotNull String newEmail, @NotNull String password);
    Observable<Void> updatePassword(@NotNull String newPassword, @NotNull String oldPassword, String oldPasswordConfirmation);

    @NotNull
    Observable<Stats> allocatePoint(@Nullable User user, @Stats.StatsTypes String s);
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/data/UserRepository.kt
package com.habitrpg.android.habitica.data

import com.habitrpg.android.habitica.models.Achievement
import com.habitrpg.android.habitica.models.QuestAchievement
import com.habitrpg.android.habitica.models.Skill
import com.habitrpg.android.habitica.models.TeamPlan
import com.habitrpg.android.habitica.models.inventory.Customization
import com.habitrpg.android.habitica.models.responses.SkillResponse
import com.habitrpg.android.habitica.models.responses.UnlockResponse
import com.habitrpg.android.habitica.models.social.Group
import com.habitrpg.android.habitica.models.tasks.Task
import com.habitrpg.android.habitica.models.user.Stats
import com.habitrpg.android.habitica.models.user.User
import com.habitrpg.android.habitica.models.user.UserQuestStatus
import com.habitrpg.shared.habitica.models.responses.VerifyUsernameResponse
import com.habitrpg.shared.habitica.models.tasks.Attribute
import kotlinx.coroutines.flow.Flow

interface UserRepository : BaseRepository {
    fun getUser(): Flow<User?>
    fun getUser(userID: String): Flow<User?>

    suspend fun updateUser(updateData: Map<String, Any?>): User?
    suspend fun updateUser(key: String, value: Any?): User?

    suspend fun retrieveUser(withTasks: Boolean = false, forced: Boolean = false, overrideExisting: Boolean = false): User?

    suspend fun revive(): User?

    suspend fun resetTutorial(): User?

    suspend fun sleep(user: User): User?

    fun getSkills(user: User): Flow<List<Skill>>

    fun getSpecialItems(user: User): Flow<List<Skill>>

    suspend fun useSkill(key: String, target: String?, taskId: String): SkillResponse?
    suspend fun useSkill(key: String, target: String?): SkillResponse?

    suspend fun disableClasses(): User?
    suspend fun changeClass(selectedClass: String? = null): User?

    suspend fun unlockPath(path: String, price: Int): UnlockResponse?
    suspend fun unlockPath(customization: Customization): UnlockResponse?

    suspend fun runCron(tasks: MutableList<Task>)
    suspend fun runCron()

    suspend fun readNotification(id: String): List<Any>?
    suspend fun readNotifications(notificationIds: Map<String, List<String>>): List<Any>?
    suspend fun seeNotifications(notificationIds: Map<String, List<String>>): List<Any>?

    suspend fun changeCustomDayStart(dayStartTime: Int): User?

    suspend fun updateLanguage(languageCode: String): User?

    suspend fun resetAccount(): User?
    suspend fun deleteAccount(password: String): Void?

    suspend fun sendPasswordResetEmail(email: String): Void?

    suspend fun updateLoginName(newLoginName: String, password: String? = null): User?
    suspend fun updateEmail(newEmail: String, password: String): Void?
    suspend fun updatePassword(oldPassword: String, newPassword: String, newPasswordConfirmation: String): Void?
    suspend fun verifyUsername(username: String): VerifyUsernameResponse?

    suspend fun allocatePoint(stat: Attribute): Stats?
    suspend fun bulkAllocatePoints(strength: Int, intelligence: Int, constitution: Int, perception: Int): Stats?

    suspend fun useCustomization(type: String, category: String?, identifier: String): User?
    suspend fun retrieveAchievements(): List<Achievement>?
    fun getAchievements(): Flow<List<Achievement>>
    fun getQuestAchievements(): Flow<List<QuestAchievement>>

    fun getUserQuestStatus(): Flow<UserQuestStatus>

    suspend fun reroll(): User?
    suspend fun retrieveTeamPlans(): List<TeamPlan>?
    fun getTeamPlans(): Flow<List<TeamPlan>>
    suspend fun retrieveTeamPlan(teamID: String): Group?
    fun getTeamPlan(teamID: String): Flow<Group?>
}
